src/Raven.Server/Documents/Indexes/Index.cs

```diff
@@ -1695,6 +1743,66 @@ public bool CanContinueBatch(IndexingStatsScope stats)
             return true;
         }
 
+        public IOperationResult Compact(Action<IOperationProgress> onProgress)
+        {
+            if (_isCompactionInProgress)
+                throw new InvalidOperationException($"Index '{Name} ({IndexId})' cannot be compacted because compaction is already in progress.");
+
+            if (_environment.Options.IncrementalBackupEnabled)
+                throw new InvalidOperationException($"Index '{Name} ({IndexId})' cannot be compacted because incremental backup is enabled.");
+
+            if (Configuration.RunInMemory)
+                throw new InvalidOperationException($"Index '{Name} ({IndexId})' cannot be compacted because it runs in memory.");
+
+            lock (_locker)
+            {
+                _isCompactionInProgress = true;
```

Comment:
This isn't thread safe.
what is going to happen to queries that are currently running? Or indexing that is currently using?
You are pulling the carpet on them by disposing the environment, and then making the memory they use invalid.

You need to add a reader / writer lock here to ensure that we are the only one accessing it.